module/VuFind/src/VuFind/RateLimiter/Turnstile/TurnstileFactory.php

```diff
@@ -0,0 +1,103 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Service
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+
+namespace VuFind\RateLimiter\Turnstile;
+
+use Laminas\ServiceManager\Exception\ServiceNotCreatedException;
+use Laminas\ServiceManager\Exception\ServiceNotFoundException;
+use Laminas\ServiceManager\Factory\FactoryInterface;
+use Psr\Container\ContainerExceptionInterface as ContainerException;
+use Psr\Container\ContainerInterface;
+use VuFind\Service\GetServiceTrait;
+
+/**
+ * Turnstile service factory.
+ *
+ * @category VuFind
+ * @package  RateLimiter
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class TurnstileFactory implements FactoryInterface
+{
+    use GetServiceTrait;
+
+    /**
+     * Create an object
+     *
+     * @param ContainerInterface $container     Service manager
+     * @param string             $requestedName Service being created
+     * @param null|array         $options       Extra options (optional)
+     *
+     * @return object
+     *
+     * @throws ServiceNotFoundException if unable to resolve the service.
+     * @throws ServiceNotCreatedException if an exception is raised when
+     * creating a service.
+     * @throws ContainerException&\Throwable if any other error occurs
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $requestedName,
+        array $options = null
+    ) {
+        if (!empty($options)) {
+            throw new \Exception('Unexpected options sent to factory.');
+        }
+
+        $this->serviceLocator = $container;
+
+        $yamlReader = $container->get(\VuFind\Config\YamlReader::class);
+        $config = $yamlReader->get('RateLimiter.yaml');
+
+        return new $requestedName(
+            $config,
+            $this->createTurnstileCache($config)
+        );
+    }
+
+    /**
+     * Create a cache for Turnstile results.
+     *
+     * @param array $config Rate limiter configuration
+     *
+     * @return ?StorageInterface
+     */
+    protected function createTurnstileCache(array $config): \Laminas\Cache\Storage\StorageInterface
+    {
+        $turnstileConfig = unserialize(serialize($config));
+        $storageOptions = $turnstileConfig['Storage']['options'] ?? [];
+        $storageOptions['namespace'] = $storageOptions['turnstileNamespace'] ?? 'Turnstile';
+        $cacheManager = $this->getService(\VuFind\Cache\Manager::class);
+        $cache = $cacheManager->getCache('object', $storageOptions['namespace']);
+        $cache->getOptions()->setTtl($storageOptions['turnstileTtl'] ?? 60 * 60 * 24);
```

Comment:
Preserving my comment from before the refactoring
> So createTurnstileCache is for the cache that records whether you have already taken a Turnstile challenge, and what the result was. Using \VuFind\CacheManager is a problem; I definitely don't want to use a file-based cache for the same performance reason that it's not used for the rest of the RateLimiter. I could use the createCache method above to create (yet) another in-memory cache instance for this purpose, but at minimum I'd have to mess with it a little to pass in the appropriate TTL, and there might be a better approach.

Reply:
In my ideal world there is a $cacheManger->getFastCache() method that usesr @EreMaijala 's createCache logic from RateLimiterManagerFactory.